### DISCUSSION TITLE
Replace New Lines w/ Regex Attachment Handler

### DISCUSSION THREAD

**Discussion by willjones9981** [nextgenhealthcare/connect#6122] [Q&A]
I have the below Regex which will turn whatever appears in my message prior to the first occurrence of "^^^" in to an attachment. 

    ^(.*?)(?:\^\^\^|$)

An issue I am running in to however is with messages where the whole message isn't on a single line such as this example below:

    This is
    a test message^^^absdefghi^^^123456

When I sent that message through it doesn't create any attachments. I have been trying to look in to what the Inbound and Outbound Replacement sections do. I tried putting \r and \n in each of them and replacing them with nothing, but that doesn't seem to be working. Am I missing something?

**Comment by Hueli** [comment]
Please try this one:
`^((?:.|\s)*?)(?:\^\^\^)`

**Comment by pacmano1** [comment]
@willjones9981 , it is sort of an odd attachment grab, mind sharing the use case? Curious.

**Reply by willjones9981** [reply]
The "This is a test message" is typically a CCD. I made it the test string for simplicity and obvious PHI reasons. Sometimes the CCD comes all on 1 line, other times it does not. If it is not all on one line then the attachment doesn't happen.

**Reply by pacmano1** [reply]
Ah, so the CDA is not the entire message?  (I do a lot of CDA work and the CDA is always the entire file).  If it is wrapped in JSON, that's a different regex of course.   And your CDA would be wrapped in ClinicalDocument tags so I would base the regex of that.

**Reply by willjones9981** [reply]
The message is a CDA with some information appended to the end of it with "^^^" being the delimiter. I just need to know if there is a way to get the whole message on a flat single line

**Reply by pacmano1** [reply]
A bit if a differeent approach, in the preprocessor:

```JS
var stuff = message.split('^^^');
var somejson = {
    "cda": stuff[0],
    "the_rest" : stuff.slice(1).join('^^^') // Joins the remaining elements back into a string
};

return JSON.stringify(somejson)
```
The regex to capture the CDA in the JSON should be easy.

**Comment by tonygermano** [comment]
`\A(?s)(.*?)(?:\^\^\^|\Z)` should work. 

- `\A` is the beginning of the input, as opposed to `^`, which is the beginning of a line.
- Likewise, `\Z` is the end of input, excluding a final terminator if it exists, where `$` is the end of a line.
- `(?s)` puts it into a mode where the `.` can match newlines, which it normally does not.

https://docs.oracle.com/javase/1.5.0/docs/api/java/util/regex/Pattern.html
